WordPress/gutenberg
# Add more taxonomy options to the post navigation link

## What?

Updated Nov 6 2023.

This PR adds support for two parameters of [get_next_post_link](https://developer.wordpress.org/reference/functions/get_next_post_link/) / [get_previous_post_link](https://developer.wordpress.org/reference/functions/get_previous_post_link/) that were not supported by the block.

$in_same_term bool Optional
Whether link should be in a same taxonomy term.
Default: false

$taxonomy string Optional
Taxonomy, if $in_same_term is true. Default 'category'.
Default: 'category'

- The feature for excluding terms was initially included, but has been removed based on feedback that it was too confusing.

Closes https://github.com/WordPress/gutenberg/issues/47090

## Why?

The features are available in classic themes but were missing from the block. The feature allows site creators and developers to limit what should be included as a next and previous link.

This allows site creators to for example add a 'news' category to some posts, and only link to the other posts in this category.

## How?

Adds a select list with taxonomies under the Advanced panel. Selecting a taxonomy in this list will limit the next and previous links.

To get a list of all supported taxonomies, including custom taxonomies, we need to know the post type.
I was not able to find any other way to get the post type besides using the core editor store.
If you know of a better way please comment on the PR.

## Testing Instructions
Create three identical posts:
Add the posts to the "birds" category, or any new category that you want to add for the test.
Add the next and previous post navigation link blocks. Open the advanced panel and select category as the taxonomy.

View the first post on the front.
You should be able to navigate to the second post in the birds category, and from there, to the third post in the birds category.
When viewing the third post, the "next" link should not show.

Register a custom post type and custom taxonomy. The fastest way to do this is to enable a plugin with custom post types, I used https://wordpress.org/plugins/yoast-test-helper/.
Repeat the testing instructions above but use the custom post type and custom taxonomy (movie and genre for example).
Only the taxonomies supported by the post type should show in the taxonomy option.

Next, with a block theme active, please test the block in the Site Editor:
Go to Appearance > Editor > Templates, and edit the single post template.
Select or add the post navigation link blocks, and enable the filter under the Advanced panel.
Save the template changes.
Confirm that the placeholder in the Site Editor is the same, and that the correct posts are linked on the front of the website.

## Screenshots or screencast 
<img width="224" alt="Filter option under the Advanced panel" src="https://github.com/WordPress/gutenberg/assets/7422055/0b8133d9-a136-4e2f-8b63-15174e5cd800">
